smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/MutinyEmitterAndAwaitInjectionTest.java

```diff
@@ -0,0 +1,675 @@
+                exceptionList.add(ce);
+            }
+            try {
+                emitter.sendAndAwait("d");
+            } catch (Exception ce) {
+                exceptionList.add(ce);
+            }
+            emitter.complete();
+        }
+
+        List<Exception> getExceptions() {
+            return exceptionList;
+        }
+
+        @Incoming("foo")
+        @Acknowledgment(Strategy.MANUAL)
+        public CompletionStage<Void> consume(final Message<String> s) {
+            list.add(s.getPayload());
+
+            if ("c".equals(s.getPayload())) {
+                return s.nack(new IllegalArgumentException("c found"));
+            } else {
+                return s.ack();
+            }
+
+        }
+    }
+
+    @ApplicationScoped
+    public static class MyMessageBeanEmittingPayloadsWithAck {
+        @Inject
+        @Channel("foo")
+        MutinyEmitter<String> emitter;
+        private final List<String> list = new CopyOnWriteArrayList<>();
+
+        public MutinyEmitter<String> emitter() {
+            return emitter;
+        }
+
+        public List<String> list() {
+            return list;
+        }
+
+        public void run() {
+            emitter.sendAndAwait(new MyMessageBean<>("a"));
+            emitter.sendAndAwait(new MyMessageBean<>("b"));
+            emitter.sendAndAwait(new MyMessageBean<>("c"));
+            emitter.complete();
+        }
+
+        @Incoming("foo")
+        public void consume(final String s) {
+            list.add(s);
+        }
+    }
+
+    public static class MyMessageBean<T> implements Message<T> {
+
+        private final T payload;
+
+        MyMessageBean(T payload) {
+            this.payload = payload;
+        }
+
+        @Override
+        public T getPayload() {
+            return payload;
+        }
+
+    }
+
+    @ApplicationScoped
+    public static class MyBeanEmittingMessages {
+        @Inject
+        @Channel("foo")
+        MutinyEmitter<String> emitter;
+        private final List<String> list = new CopyOnWriteArrayList<>();
+
+        public MutinyEmitter<String> emitter() {
+            return emitter;
+        }
+
+        public List<String> list() {
+            return list;
+        }
+
+        public void run() {
+            emitter.sendAndAwait(Message.of("a"));
+            emitter.sendAndAwait(Message.of("b"));
+            emitter.sendAndAwait(Message.of("c"));
+
+        }
+
+        @Incoming("foo")
+        public void consume(final String s) {
+            list.add(s);
+        }
+    }
+
+    @ApplicationScoped
+    public static class MyBeanEmittingMessagesUsingStream {
+        @Inject
+        @Channel("foo")
+        MutinyEmitter<String> emitter;
+        private final List<String> list = new CopyOnWriteArrayList<>();
+
+        public MutinyEmitter<String> emitter() {
+            return emitter;
+        }
+
+        public List<String> list() {
+            return list;
+        }
+
+        public void run() {
```

Comment:
Try running on another thread, as await is blocking the caller thread until acknowledgment.

Reply:
Also, having a thread dump to verify where the blocking happens would be great.

Reply:
So `run()` on a separate thread? Can give that a go

Reply:
yes